### PR TITLE
[L2B-5989] Add liveness aggregating and anomalies indexers

### PR DESCRIPTION
**Pull request by maciekop-l2b**
Resolves L2B-5989

### CODE REVIEW

**Review by maciekzygmunt — CHANGES_REQUESTED** [review]
You need to update the Prisma schema AND update changed repositories in database package. We need to maintain both of them before we make a full migration.

**Comment by maciekop-l2b** [comment]
> You need to update the Prisma schema AND update changed repositories in database package. We need to maintain both of them before we make a full migration.

Done

File: packages/backend/src/modules/tracked-txs/modules/liveness/indexers/AnomaliesIndexer.test.ts [hunk 1]
```diff
@@ -0,0 +1,331 @@
...
+      expect(mockCalculateAnomalies).toHaveBeenCalledWith(NOW.toStartOf('day'))
+
+      expect(mockAnomaliesRepository.addOrUpdateMany).toHaveBeenCalledWith(
+        mockAnomalies,
+      )
+
+      expect(result).toEqual(NOW.toStartOf('day').toNumber())
+    })
+  })
+
```
Comment (torztomasz) [inline packages/backend/src/modules/tracked-txs/modules/liveness/indexers/AnomaliesIndexer.test.ts:105]: aren't we missing test that just updates without adjusting?

File: packages/backend/src/modules/tracked-txs/modules/liveness/indexers/AnomaliesIndexer.ts [hunk 2]
```diff
@@ -0,0 +1,265 @@
...
+
+      this.logger.debug('Liveness records loaded', {
+        projectId: project.projectId,
+        count: livenessRecords.length,
+      })
+
+      const [batchSubmissions, stateUpdates, proofSubmissions] =
+        groupByType(livenessRecords)
+
+      const test = this.detectAnomalies(
```
Comment (torztomasz) [inline packages/backend/src/modules/tracked-txs/modules/liveness/indexers/AnomaliesIndexer.ts:109]: I guess it should be named differently or just used inline inside a fn

File: packages/backend/src/modules/tracked-txs/modules/liveness/indexers/AnomaliesIndexer.ts [hunk 3]
```diff
@@ -0,0 +1,265 @@
...
+      subtype,
+    })
+
+    const intervals = calculateIntervals(livenessRecords)
+
+    const lastIndex = intervals.findIndex((interval) =>
+      interval.record.timestamp.lte(to.add(-1 * this.SYNC_RANGE, 'days')),
+    )
+
+    const { means, stdDevs } = this.calculate30DayRollingStats(
```
Comment (torztomasz) [inline packages/backend/src/modules/tracked-txs/modules/liveness/indexers/AnomaliesIndexer.ts:171]: whats stdDevs?

File: packages/backend/src/modules/tracked-txs/modules/liveness/indexers/AnomaliesIndexer.ts (hunk 2 above)
Comment (maciekop-l2b) [reply packages/backend/src/modules/tracked-txs/modules/liveness/indexers/AnomaliesIndexer.ts:109]: argh... leftover :)

File: packages/backend/src/modules/tracked-txs/modules/liveness/indexers/AnomaliesIndexer.ts (hunk 3 above)
Comment (maciekop-l2b) [reply packages/backend/src/modules/tracked-txs/modules/liveness/indexers/AnomaliesIndexer.ts:171]: Renamed it to `stdDeviations` to be more readable

File: packages/backend/src/modules/tracked-txs/modules/liveness/indexers/AnomaliesIndexer.test.ts (hunk 1 above)
Comment (maciekop-l2b) [reply packages/backend/src/modules/tracked-txs/modules/liveness/indexers/AnomaliesIndexer.test.ts:105]: Added

File: packages/backend/src/modules/tracked-txs/modules/liveness/indexers/AnomaliesIndexer.ts [hunk 4]
```diff
@@ -0,0 +1,265 @@
...
+  extends Omit<ManagedChildIndexerOptions, 'name'> {
+  livenessRepository: LivenessRepository
+  anomaliesRepository: AnomaliesRepository
+  projects: Project[]
+}
+
+export class AnomaliesIndexer extends ManagedChildIndexer {
+  private readonly SYNC_RANGE = 30
+  constructor(private readonly $: AnomaliesIndexerIndexerDeps) {
+    super({ ...$, name: 'anomalies' })
```
Comment (antooni) [inline packages/backend/src/modules/tracked-txs/modules/liveness/indexers/AnomaliesIndexer.ts:38]: anomalies it too generic name, can lead to overlap in the future

File: packages/backend/src/modules/tracked-txs/modules/liveness/repositories/AggregatedLivenessRepository.ts [hunk 5]
```diff
@@ -0,0 +1,76 @@
...
+
+export type AggregatedLivenessRange = '30D' | '90D' | 'MAX'
+
+export interface AggregatedLivenessRecord {
+  projectId: string
+  subtype: TrackedTxsConfigSubtype
+  range: AggregatedLivenessRange
+  min: number
+  avg: number
+  max: number
```
Comment (antooni) [inline packages/backend/src/modules/tracked-txs/modules/liveness/repositories/AggregatedLivenessRepository.ts:18]: rename to updated_at

File: packages/backend/src/modules/tracked-txs/modules/liveness/indexers/LivenessAggregatingIndexer.ts [hunk 6]
```diff
@@ -0,0 +1,239 @@
...
+    safeHeight: number,
+    parentSafeHeight: number,
+  ): Promise<number> {
+    const now = UnixTime.now()
+    const endOfPeviousDay = now.toStartOf('day').add(-1, 'seconds')
+    let targetHeight = new UnixTime(safeHeight)
+      .toEndOf('day')
+      .add(-1, 'seconds')
+
+    if (parentSafeHeight <= endOfPeviousDay.toNumber()) {
```
Comment (antooni) [inline packages/backend/src/modules/tracked-txs/modules/liveness/indexers/LivenessAggregatingIndexer.ts:54]: typo

**Review by antooni — APPROVED** [review]

File: packages/backend/src/modules/tracked-txs/modules/liveness/indexers/AnomaliesIndexer.ts (hunk 4 above)
Comment (maciekop-l2b) [reply packages/backend/src/modules/tracked-txs/modules/liveness/indexers/AnomaliesIndexer.ts:38]: Let's leave it as it is for know as this is exactly the type of object we are dealing with

File: packages/backend/src/modules/tracked-txs/modules/liveness/indexers/LivenessAggregatingIndexer.ts (hunk 6 above)
Comment (maciekop-l2b) [reply packages/backend/src/modules/tracked-txs/modules/liveness/indexers/LivenessAggregatingIndexer.ts:54]: Fixed

File: packages/backend/src/modules/tracked-txs/modules/liveness/repositories/AggregatedLivenessRepository.ts (hunk 5 above)
Comment (maciekop-l2b) [reply packages/backend/src/modules/tracked-txs/modules/liveness/repositories/AggregatedLivenessRepository.ts:18]: Done

**Pull request merged**